### ISSUE TITLE
Recursive type in match arms fails to narrow

### ISSUE THREAD

**Issue by anuraaga** [astral-sh/ty#3983]
### Summary

When updating ty, I found code that checked before fails now (pyright is fine).

```python
from collections.abc import Mapping, Sequence
from typing import TypeAlias


RecursiveValue: TypeAlias = (
    None
    | bool
    | int
    | float
    | str
    | Sequence["RecursiveValue"]
    | Mapping[str, "RecursiveValue"]
)


def accepts_mapping(value: Mapping[str, RecursiveValue]) -> None:
    pass


def accepts_value(value: RecursiveValue) -> None:
    match value:
        case None:
            pass
        case bool():
            pass
        case int() | float():
            pass
        case str():
            pass
        case Mapping() as mapping:
            accepts_mapping(mapping)
        case Sequence() as sequence:
            pass

❯ uv run ty check .repros/ty_recursive_mapping_match.py
error[[invalid-argument-type](https://ty.dev/rules#invalid-argument-type)]: Argument to function `accepts_mapping` is incorrect
  --> .repros/ty_recursive_mapping_match.py:31:29
   |
31 |             accepts_mapping(mapping)
   |                             ^^^^^^^ Expected `Mapping[str, None | int | float | ... omitted 3 union elements]`, found `(Sequence[Divergent] & Top[Mapping[Unknown, object]] & ~int & ~float & ~str) | (Mapping[str, Divergent] & ~int & ~float & ~str)`
   |
info: element `Sequence[Divergent] & Top[Mapping[Unknown, object]] & ~int & ~float & ~str` of union `(Sequence[Divergent] & Top[Mapping[Unknown, object]] & ~int & ~float & ~str) | (Mapping[str, Divergent] & ~int & ~float & ~str)` is not assignable to `Mapping[str, None | int | float | ... omitted 3 union elements]`
info: Function defined here
  --> .repros/ty_recursive_mapping_match.py:16:5
   |
16 | def accepts_mapping(value: Mapping[str, RecursiveValue]) -> None:
   |     ^^^^^^^^^^^^^^^ ----------------------------------- Parameter declared here
   |

Found 1 diagnostic
```

I could confirm, this started happening with 0.0.53 and 0.0.52 works. It is presumably related to some of the recursion fixes.

### Version

ty 0.0.59 (71bdf3104 2026-07-12)

**Comment by charliermarsh** [comment]
Strictly speaking, this is correct, because you can have a subclass like this:
```python
class Both(Sequence[str], Mapping[str, bytes]):
    @overload
    def __getitem__(self, key: int) -> str: ...

    @overload
    def __getitem__(self, key: slice) -> Sequence[str]: ...

    @overload
    def __getitem__(self, key: str) -> bytes: ...

    def __getitem__(
        self, key: int | slice | str
    ) -> str | Sequence[str] | bytes:
        if isinstance(key, str):
            return b"not a RecursiveValue"
        if isinstance(key, slice):
            return ()
        return "sequence item"

    def __iter__(self) -> Iterator[str]:
        return iter(("key",))

    def __len__(self) -> int:
        return 1
```

Here, `value: RecursiveValue = Both()` is accepted via `Sequence[str]`. But when you hit `case Mapping() as mapping`, it's `Mapping[str, bytes]`, not `Mapping[str, RecursiveValue]`.

We're considering a few ways to allow "unsound" narrowing like this by default (with opt-in for the stricter, correct behavior we enforce here).

**Comment by carljm** [comment]
#1578 is our general tracking issue for this strict handling of possible multiple inheritance in narrowing unions. Closing this as duplicate of that.
